docs/source-2.0/guides/client-guidance/application-protocols/http.md

```diff
@@ -0,0 +1,179 @@
+metadata. As of {rfc}`9110`, these key/value pairs are exclusively referred to
+as {rfc}`fields <9110#section-5>`.
+
+When designing HTTP interfaces for Smithy clients, be careful to understand
+field semantics. In particular, it is important to understand that field keys
+are case-insensitive and may appear more than once in an HTTP message. Since
+field keys may appear more than once, it is recommended that they are
+represented as an iterable collection of key/value pairs or a map whose value
+type is a list. This allows protocol implementations to handle joining and
+splitting.
+
+```java
+public interface HttpFields extends Iterable<Map.Entry<String, List<String>>> {
+    /**
+     * Create an HttpFields instance from a map.
+     *
+     * @param fields Field map to use as a data source.
+     * @return the created fields.
+     */
+    static HttpFields of(Map<String, List<String>> fields) {
+        // This constructs a theoretical default implementation of the
+        // HttpFields interface that creates an unmodifiable copy of the given
+        // map.
+        return fields.isEmpty() ? UnmodifiableHttpFields.EMPTY : new UnmodifiableHttpFields(fields);
+    }
+
+    /**
+     * Convert the HttpFields to a map.
+     *
+     * @return the fields as a map.
+     */
+    Map<String, List<String>> toMap();
+
+    /**
+     * Check if the given field is case-insensitively present.
+     *
+     * @param name Name of the field to check.
+     * @return true if the field is present.
+     */
+    default boolean containsField(String name) {
+        return !getAllValues(name).isEmpty();
+    }
+
+    /**
+     * Get the first field value of a specific field by case-insensitive name.
+     * 
+     * Smithy clients know whether a given header should have a single value or
+     * a list value. This helper method simplifies usage for headers with a
+     * single value.
+     *
+     * @param name Name of the field to get.
+     * @return the matching field value, or null if not found.
+     */
+    default String getFirstValue(String name) {
+        var list = getAllValues(name);
+        return list.isEmpty() ? null : list.get(0);
+    }
+
+    /**
+     * Get the values of a specific field by case-insensitive name.
+     *
+     * @param name Name of the field to get the values of.
+     * @return the values of the field, or an empty list.
+     */
+    List<String> getAllValues(String name);
+}
+```
+
+#### Implementation recommendations
+
+It is not recommended to automatically attempt to join values for a given field
+key at the HTTP layer. {rfc}`9110#section-5` allows field values to be joined
+with a comma, but doing so automatically can introduce data corruption if one of
+the field values already includes a comma. {rfc}`Section 5.6 <9110#section-5.6>`
+includes productions that can help to handle those edge cases, but whether they
+are used or not is up to the protocol definition.
```

Comment:
What should implementors do instead? Are we implying that all HTTP client implementations will be able to _accept_ lists of header values (vs merely a single value)?